File: frontend/src/types/knowledge.ts
export type KnowledgeFile = {
  id: number;
  filename: string;
  content_text: string;
  status: string;
  created_at: string;
};

export type KnowledgeChunk = {
  id: number;
  file_id: number;
  chunk_index: number;
  content: string;
  metadata_json: Record<string, unknown>;
  created_at: string;
};

export type RetrievedChunk = {
  chunk_id: number;
  file_id: number;
  content: string;
  score: number;
  metadata: Record<string, unknown>;
  retriever_type: string;
};

File: backend/app/models/knowledge.py
from datetime import datetime

from sqlalchemy import JSON, DateTime, ForeignKey, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base
from app.models.timestamps import utc_now


class KnowledgeFile(Base):
    __tablename__ = "knowledge_files"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    filename: Mapped[str] = mapped_column(String(255), nullable=False)
    content_text: Mapped[str] = mapped_column(Text, nullable=False)
    status: Mapped[str] = mapped_column(String(50), nullable=False, default="uploaded")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now, nullable=False)

    chunks: Mapped[list["KnowledgeChunk"]] = relationship(
        back_populates="file",
        cascade="all, delete-orphan",
    )


class KnowledgeChunk(Base):
    __tablename__ = "knowledge_chunks"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    file_id: Mapped[int] = mapped_column(ForeignKey("knowledge_files.id"), index=True, nullable=False)
    chunk_index: Mapped[int] = mapped_column(nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    metadata_json: Mapped[dict] = mapped_column(JSON, nullable=False, default=dict)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now, nullable=False)

    file: Mapped[KnowledgeFile] = relationship(back_populates="chunks")
